### ISSUE TITLE
[feature request] Support tensors of different sizes as batch elements in DataLoader

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#1512]
Motivating example is returning bounding box annotation for images along with an image. An annotation list can contain variable number of boxes depending on an image, and padding them to a single length (and storing that length) may be nasty and unnecessarily complex.

```python
import torch.utils.data
loader = torch.utils.data.DataLoader(dataset = [(torch.zeros(3, 128, 128), torch.zeros(i, 4)) for i in range(1, 3)], batch_size = 2)
for batch in loader:
    print(batch)
```

Currently this blows with a message below because collate wants to `torch.stack` batch elements, regardless if they have same size:
```
File "...torch/utils/data/dataloader.py", line 188, in __next__
    batch = self.collate_fn([self.dataset[i] for i in indices])
  File ".../torch/utils/data/dataloader.py", line 110, in default_collate
    return [default_collate(samples) for samples in transposed]
  File ".../torch/utils/data/dataloader.py", line 92, in default_collate
    return torch.stack(batch, 0, out=out)
  File ".../torch/functional.py", line 56, in stack
    inputs = [t.unsqueeze(dim) for t in sequence]
RuntimeError: cannot unsqueeze empty tensor at .../torch/lib/TH/generic/THTensor.c:530
```

Returning a list instead of variable-sized tensor doesn't work either. Providing a custom collate isn't very nice either, since most of the default behavior needs to be copied, and the default collate doesn't allow hooks.

A solution would be either adding an easy way to extend the default collate, or changing the first collate's branch to something like:

```python
if all(map(torch.is_tensor, batch)) and any([tensor.size() != batch[0].size() for tensor in batch]):
     return batch
```

As a workaround, I'm currently monkey-patching the default collate like this:
```python
collate_old = torch.utils.data.dataloader.default_collate
torch.utils.data.dataloader.default_collate = lambda batch: batch if all(map(torch.is_tensor, batch)) and any([tensor.size() != batch[0].size() for tensor in batch]) else collate_old(batch)
```

**Comment by apaszke** [comment]
I don't think it's a good idea to add that to the default collate_fn. It sounds like an application specific thing. My solution would look like below. Seems quite simple and non-hacky:
```python
from torch.utils.data.dataloader import default_collate

class BBoxList(list): pass

class MyDataset(Dataset):
    def __getitem__(self, i):
        return self._load_sample(i), BBoxList([self._load_bbox(i)])

def collate_fn(batch):
    if isinstance(batch, BBoxList):
        return batch
    return default_collate(batch)

dataloader = torch.utils.data.DataLoader(..., collate_fn=collate_fn)
```

**Comment by vadimkantorov** [comment]
I don't consider function composition a monkey-patching either. The problem is I failed to find a nice function composition solution and had to redefine `default_collate`.

I tried your proposed way like this (please correct me if I got it wrong!):
```python
import torch.utils.data
from torch.utils.data.dataloader import default_collate

class BBoxList(list): pass

def collate_fn(batch):
    if isinstance(batch, BBoxList):
        return batch
    return default_collate(batch)

loader = torch.utils.data.DataLoader(dataset = [(torch.zeros(2, 2), BBoxList(list(torch.zeros(i, 4)))) for i in range(3, 5)], batch_size = 2, collate_fn = collate_fn)

# these don't work either:
# loader = torch.utils.data.DataLoader(dataset = [(torch.zeros(2, 2), BBoxList([list(torch.zeros(i, 4))])) for i in range(3, 5)], batch_size = 2, collate_fn = collate_fn)
# loader = torch.utils.data.DataLoader(dataset = [(torch.zeros(2, 2), BBoxList([torch.zeros(i, 4)])) for i in range(3, 5)], batch_size = 2, collate_fn = collate_fn)

for batch in loader:
    print(batch)
```

The first example will have a BBoxList with 3  bboxes, the second example will have a BBoxList with 4 bboxes.

on my machine this prints:
```
[
(0 ,.,.) =
  0  0
  0  0

(1 ,.,.) =
  0  0
  0  0
[torch.FloatTensor of size 2x2x2]
,  [
 0  0
 0  0
[torch.FloatTensor of size 2x2]
,
 0  0
 0  0
[torch.FloatTensor of size 2x2]
,
 0  0
 0  0
[torch.FloatTensor of size 2x2]
]]
```

The first element of batch is returned correctly: two stacked images.  The second is incorrect. It should have been an array of two elements with two tensors inside (of dims 3x2 and 4x2), the 4 bboxes seems sort of lost.

**Comment by apaszke** [comment]
Ok, I now see what's the problem. I'll try to think how to best solve it

**Comment by mnicnc404** [comment]
Please correct me if I get it wrong, but I think `if isinstance(batch, BBoxList):` should be `if isinstance(batch[0][1], BBoxList):`, depending on the order of the output of `__getitem__`, since the type of `batch` is list but not BBoxList.

Following the code of @vadimkantorov, It should be:

```
import torch.utils.data
from torch.utils.data.dataloader import default_collate

class BBoxList(list): pass

def collate_fn(batch):
    ###################
    if isinstance(batch[0][1], BBoxList):
    ###################
        return batch
    return default_collate(batch)

loader = torch.utils.data.DataLoader(dataset = [(torch.zeros(2, 2), BBoxList(list(torch.zeros(i, 4)))) for i in range(3, 5)], batch_size = 2, collate_fn = collate_fn)

for batch in loader:
    print(batch)
```

So the condition is fulfilled. On my machine, this prints:

```
[(
 0  0
 0  0
[torch.FloatTensor of size 2x2]
, [
 0
 0
 0
 0
[torch.FloatTensor of size 4]
,
 0
 0
 0
 0
[torch.FloatTensor of size 4]
,
 0
 0
 0
 0
[torch.FloatTensor of size 4]
]), (
 0  0
 0  0
[torch.FloatTensor of size 2x2]
, [
 0
 0
 0
 0
[torch.FloatTensor of size 4]
,
 0
 0
 0
 0
[torch.FloatTensor of size 4]
,
 0
 0
 0
 0
[torch.FloatTensor of size 4]
,
 0
 0
 0
 0
[torch.FloatTensor of size 4]
])]
```
And I think this is what @vadimkantorov wants.

**Comment by mnicnc404** [comment]
Please correct me if I get it wrong, but `isinstance(batch, BBoxList)` would never be evoked since `batch` is a list.

An alternative, way which might be not so efficient (following the settings from @vadimkantorov):
```
import torch
from torch.utils.data import DataLoader
# No need to define a new class
# Suppose you know the order of your customized Dataset

def collate_fn(batch):
    # Note that batch is a list 
    batch = list(map(list, zip(*batch)))  # transpose list of list
    out = None
    # You should know that batch[0] is a fixed-size tensor since you're using your customized Dataset
    # reshape batch[0] as (N, H, W)
    # batch[1] contains tensors of different sizes; just let it be a list.
    # If your num_workers in DataLoader is bigger than 0
    #     numel = sum([x.numel() for x in batch[0]])
    #     storage = batch[0][0].storage()._new_shared(numel)
    #     out = batch[0][0].new(storage)
    batch[0] = torch.stack(batch[0], 0, out=out)
    return batch

loader = DataLoader(dataset = [(torch.zeros(2, 2), torch.zeros(i, 4)) for i in range(3, 5)], batch_size = 2, collate_fn = collate_fn)
# no need to wrap torch.zeros(i, 4) as a list
for batch in loader:
    print(batch)
for batch in loader:
    print(batch)
    print('tensor size: ', batch[0].size())
    print('bbox length: ', len(batch[1]))
    print('fisrt bbox size: ', batch[1][0].size())
    print('second bbox size: ', batch[1][1].size())
```

On my machine, this prints:
```
[
(0 ,.,.) =
  0  0
  0  0

(1 ,.,.) =
  0  0
  0  0
[torch.FloatTensor of size 2x2x2]
, [[
 0
 0
 0
 0
[torch.FloatTensor of size 4]
,
 0
 0
 0
 0
[torch.FloatTensor of size 4]
,
 0
 0
 0
 0
[torch.FloatTensor of size 4]
], [
 0
 0
 0
 0
[torch.FloatTensor of size 4]
,
 0
 0
 0
 0
[torch.FloatTensor of size 4]
,
 0
 0
 0
 0
[torch.FloatTensor of size 4]
,
 0
 0
 0
 0
[torch.FloatTensor of size 4]
]]]
tensor size:  torch.Size([2, 2, 2])
bbox length:  2
fisrt bbox size:  torch.Size([3, 4])
second bbox size:  torch.Size([4, 4])
```
I think this is what @vadimkantorov want (at least the output is).

**Comment by drcege** [comment]
@mnicnc404  But then how should I convert this tensor list (`batch[1]`) to a Variable?  Or should I pass this list directly to forward?

**Comment by jdhao** [comment]
@vadimkantorov , do you have any progress on implementation of a batch with variable  size input image?

**Comment by vadimkantorov** [comment]
@jdhao I'm using the following snippet for now:
```
torch.utils.data.dataloader.default_collate = (lambda default_collate = torch.utils.data.dataloader.default_collate: lambda batch: batch if all(map(torch.is_tensor, batch)) and any([tensor.size() != batch[0].size() for tensor in batch]) else default_collate(batch))()
```

Though, to use variable size input images, I'd suggest zero-padding them appropriately and not resorting to this.

@drcege It's an valid question, I'm currently converting every list element to a Variable, and then passing the list to forward

**Comment by jdhao** [comment]
I also thought of converting the image in a batch to a list of tensors. But then we have to feed the image to the network one by one instead of by a batch, right ? Will this lead to low processing speed?

**Comment by vadimkantorov** [comment]
@jdhao If you're dealing with images and putting them through the network, use a single padded tensor, not list.

The list was important for me for some very small tensors, like ground truth bounding boxes.

**Comment by jdhao** [comment]
@vadimkantorov, the images are of different size and aspect ratio, while it is possible to pad zeros to images to make them the same size, the network may waste a lot of time convolving with zeros.

**Comment by vadimkantorov** [comment]
@jdhao you may sample images for the minibatch in aspect-ratio-respecting way

**Comment by snakers4** [comment]
**Hi guys, I faced a similar issue and found an elegant and ingenious solution!**

What if I have a multi-head model (in my case it is **several image sizes and aspect ratios**), and I would like to learn shared representations, but do not want to iterate over several datasets (in my case it caused sequential over-fitting and poorer performance)?
Yeah, ConcatDataset does not help either in this case (it can work w/o shuffling and if your dataset exactly divides by batch-size, but it's not a proper solution).

**You see, as far a I understand, PyTorch Dataset/Dataloader abstractions work the following way:**
- First you define an abstract [Dataset ](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataset.py) with `len()` and `getitem()` methods;
- Then you define a sampling [strategy ](https://pytorch.org/docs/stable/_modules/torch/utils/data/sampler.html)(usually people just opt for shuffle=True in 95% of cases);
- Then you just use this strategy with the [Dataloader](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py);

You see, the gist is contained in this [portion ](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py#L465-L491)of the Dataloader code!

Translated in plain English it goes something like this:
- Has user re-defined standard batch-producing class? No?
- Has use re-defined standard sampling class? No?
- Then just shuffle randomly (or not) and use standard batch sampler

If you read the standard batch [sampler](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/sampler.py#L136-L144) - you will see that it just sequentially samples from a sampling straregy and yields batches when they are full. As easy as that. So ... we can just re-write RandomSampler class a bit to support our multi-cluster behaviour and we are good to go =)

**My final logic is the following:**
- Somewhere in your Dataset class define cluster_indices property which says which index refers to which cluster of data
- Chunkify your data manually `batches = [cluster_indices[i:i + self.batch_size] for i in range(0, len(cluster_indices), self.batch_size)]`
- Drop the batches that are smaller than your batch size
- Concatenate all of your cluster batch lists, flatter the list, shuffle the batches
- Flatten list once again
- Feed to the standard BatchSampler and observe the multi-threaded goodness of PyTorch Dataloader  =)

**New class**
`
import random
from torch.utils.data.sampler import Sampler

class ClusterRandomSampler(Sampler):
    r"""Takes a dataset with cluster_indices property, cuts it into batch-sized chunks
    Drops the extra items, not fitting into exact batches
    Arguments:
        data_source (Dataset): a Dataset to sample from. Should have a cluster_indices property
        batch_size (int): a batch size that you would like to use later with Dataloader class
        shuffle (bool): whether to shuffle the data or not
    """

    def __init__(self, data_source, batch_size, shuffle=True):
        self.data_source = data_source
        self.batch_size = batch_size
        self.shuffle = shuffle
        
    def flatten_list(self, lst):
        return [item for sublist in lst for item in sublist]

    def __iter__(self):

        batch_lists = []
        for cluster_indices in train_dataset.cluster_indices:
            batches = [cluster_indices[i:i + self.batch_size] for i in range(0, len(cluster_indices), self.batch_size)]
            # filter our the shorter batches
            batches = [_ for _ in batches if len(_) == self.batch_size]
            if self.shuffle:
                random.shuffle(batches)
            batch_lists.append(batches)       
        
        # flatten lists and shuffle the batches if necessary
        # this works on batch level
        lst = self.flatten_list(batch_lists)
        if self.shuffle:
            random.shuffle(lst)
        # final flatten  - produce flat list of indexes
        lst = self.flatten_list(lst)        
        return iter(lst)

    def __len__(self):
        return len(self.data_source)
`

**Usage**
`
sampler = ClusterRandomSampler(train_dataset,512,True)

train_loader = torch.utils.data.DataLoader(
    train_dataset,
    batch_size=512,        
    sampler=sampler,
    shuffle=False,
    num_workers=1,
    pin_memory=False,
    drop_last=False)
`

**Comment by Jasperty** [comment]
> **Hi guys, I faced a similar issue and found an elegant and ingenious solution!**
> 
what is the format of train_dataset?

**Comment by vadimkantorov** [comment]
Propose to reconsider this and collate list of tensors as a list if they cannot be stacked because of different sizes

**Comment by addisonklinke** [comment]
I agree that list of tensors should be supported for variable number of annotations. This is too common in object detection to not have some predefined `collate_fn` in `torchvision`

**Comment by vadimkantorov** [comment]
@cpuhrsch btw yep, creating a nestedtensor could become a default collate behavior if tensors have different sizes

**Comment by cpuhrsch** [comment]
@cpuhrsch - a NestedTensor will need to copy to be constructed from a list of Tensors, unless we introduce some kind of special layout, since torch Tensors have a single piece of underlying Storage.

**Comment by vadimkantorov** [comment]
> @cpuhrsch - a NestedTensor will need to copy to be constructed from a list of Tensors, unless we introduce some kind of special layout, since torch Tensors have a single piece of underlying Storage.

for that a TensorList abstraction would be good :)
